spec/unit/application/ssl_spec.rb

```diff
@@ -0,0 +1,284 @@
+require 'spec_helper'
+require 'puppet/application/ssl'
+require 'webmock/rspec'
+require 'openssl'
+
+describe Puppet::Application::Ssl do
+  let(:ssl) { Puppet::Application[:ssl] }
+  let(:name) { 'ssl-client' }
+
+  def generate_cert(name, issuer = nil, issuer_key = nil)
+    # generate CA key pair
+    private_key = OpenSSL::PKey::RSA.new(512)
+    public_key = private_key.public_key
+
+    # generate CSR
+    csr = OpenSSL::X509::Request.new
+    csr.version = 0
+    csr.subject = OpenSSL::X509::Name.new([["CN", name]])
+    csr.public_key = public_key
+    csr.sign(private_key, OpenSSL::Digest::SHA256.new)
+
+    # is it self-signed?
+    issuer ||= csr.subject
+    issuer_key ||= private_key
+
+    # issue cert
+    cert = OpenSSL::X509::Certificate.new
+    cert.version    = 2 # X509v3
+    cert.subject    = csr.subject
+    cert.issuer     = issuer
+    cert.public_key = csr.public_key
+    cert.serial     = 1
+    cert.not_before = Time.now - (60*60*24)
+    cert.not_after  = Time.now + (60*60*24)
+    cert.sign(issuer_key, OpenSSL::Digest::SHA256.new)
+
+    {:private_key => private_key, :csr => csr, :cert => cert}
+  end
+
+  def generate_crl(name, issuer, issuer_key)
+    crl = OpenSSL::X509::CRL.new
+    crl.version = 1
+    crl.issuer = issuer
+    crl.last_update = Time.now - (60*60*24)
+    crl.next_update =  Time.now + (60*60*24)
+    crl.extensions = [OpenSSL::X509::Extension.new('crlNumber', OpenSSL::ASN1::Integer(0))]
+    crl.sign(issuer_key, OpenSSL::Digest::SHA256.new)
+
+    crl
+  end
+
+  before :all do
+    @ca = generate_cert('ca')
+    @crl = generate_crl('ca', @ca[:cert].subject, @ca[:private_key])
+    @host = generate_cert('ssl-client', @ca[:cert].subject, @ca[:private_key])
+  end
+
+  before do
+    WebMock.disable_net_connect!
+
+    Puppet.settings.use(:main)
+    Puppet[:certname] = name
+
+    # Host assumes ca cert and crl are present
+    File.open(Puppet[:localcacert], 'w') { |f| f.write(@ca[:cert].to_pem) }
+    File.open(Puppet[:hostcrl], 'w') { |f| f.write(@crl.to_pem) }
+
+    # Setup our ssl client
+    File.open(Puppet[:hostprivkey], 'w') { |f| f.write(@host[:private_key].to_pem) }
+    File.open(Puppet[:hostpubkey], 'w') { |f| f.write(@host[:private_key].public_key.to_pem) }
+  end
+
+  def expects_command_to_output(expected_message = nil, code = 0)
+    expect {
+      expect {
+        ssl.run_command
+      }.to exit_with(code)
+    }.to output(expected_message).to_stdout
+  end
+
+  shared_examples_for 'an ssl action' do
+    it 'downloads the CA bundle first when missing' do
+      File.delete(Puppet[:localcacert])
+      stub_request(:get, %r{puppet-ca/v1/certificate/ca}).to_return(status: 200, body: @ca[:cert].to_pem)
```

Comment:
Is this something built into webmock? I should have been using this in some other tests...

Reply:
Yeah, it's super handy and automagically works with different http implementations. The main trick is to disable requests via `WebMock.disable_net_connect!` in `before each`